### PR TITLE
Improve DETR models

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR: 

- [x] fixes Deformable DETR's loss function
- [x] adds more copied from statements for consistency
- [x] fixes Conditional DETR's integration tests.

As pointed out in #18948, Deformable DETR uses the same loss function and Hungarian matcher as Conditional DETR (use of sigmoid instead of softmax and not including the no-object class).

This PR also improves the  (original; conditional; deformable) DETR models by improving docs, adding more Copied from statements.

### CODE REVIEW

File: src/transformers/models/conditional_detr/modeling_conditional_detr.py [hunk 1]
```diff
@@ -2595,9 +2592,18 @@ def generalized_box_iou(boxes1, boxes2):
...
 
 
-# below: taken from https://github.com/facebookresearch/detr/blob/master/util/misc.py#L306
+# Copied from transformers.models.detr.modeling_detr.center_to_corners_format
+def center_to_corners_format(x):
+    """
+    Converts a PyTorch tensor of bounding boxes of center format (center_x, center_y, width, height) to corners format
+    (x_0, y_0, x_1, y_1).
+    """
+    x_c, y_c, w, h = x.unbind(-1)
```
Comment (sgugger) [inline src/transformers/models/conditional_detr/modeling_conditional_detr.py:2601]: I would also upgrade the variable names here (since that's the whole goal of the PR :-) )

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 2]
```diff
@@ -2189,20 +2210,22 @@ def get_loss(self, loss, outputs, targets, indices, num_boxes):
...
 
-        Parameters:
-             outputs: dict of tensors, see the output specification of the model for the format
-             targets: list of dicts, such that len(targets) == batch_size.
-                      The expected keys in each dict depends on the losses applied, see each loss' doc
+        Args:
+             outputs (`dict`, *optional*):
+                Dictionary of tensors, see the output specification of the model for the format.
+             targets (`List[dict]`, *optional*):
+                List of dicts, such that len(targets) == batch_size. The expected keys in each dict depends on the
```
Comment (sgugger) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:2227]:
```suggestion
                List of dicts, such that `len(targets) == batch_size`. The expected keys in each dict depends on the
```
Nit

File: src/transformers/models/detr/modeling_detr.py [hunk 3]
```diff
@@ -2288,6 +2284,17 @@ def generalized_box_iou(boxes1, boxes2):
     return iou - (area - union) / area
 
 
+# Copied from transformers.models.detr.feature_extraction_detr.center_to_corners_format
+def center_to_corners_format(x):
+    """
+    Converts a PyTorch tensor of bounding boxes of center format (center_x, center_y, width, height) to corners format
+    (x_0, y_0, x_1, y_1).
+    """
+    x_c, y_c, w, h = x.unbind(-1)
```
Comment (sgugger) [inline src/transformers/models/detr/modeling_detr.py:2293]: As mentioned above, let's also change `feature_extraction_detr.center_to_corners_format` to upgrade the variable names :-)

**Review by sgugger — APPROVED** [review]
Very nice! Thanks a lot for improving these models!

File: src/transformers/models/conditional_detr/modeling_conditional_detr.py [hunk 4]
```diff
@@ -2385,7 +2385,7 @@ def forward(self, outputs, targets):
              outputs (`dict`, *optional*):
                 Dictionary of tensors, see the output specification of the model for the format.
              targets (`List[dict]`, *optional*):
-                List of dicts, such that len(targets) == batch_size. The expected keys in each dict depends on the
+                List of dicts, such that `len(targets) == batch_size. The expected keys in each dict depends on the
```
Comment (alaradirik) [inline src/transformers/models/conditional_detr/modeling_conditional_detr.py:2388]:
```suggestion
                List of dicts, such that `len(targets) == batch_size`. The expected keys in each dict depends on the
```

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 5]
```diff
@@ -2224,7 +2224,7 @@ def forward(self, outputs, targets):
              outputs (`dict`, *optional*):
                 Dictionary of tensors, see the output specification of the model for the format.
              targets (`List[dict]`, *optional*):
-                List of dicts, such that len(targets) == batch_size. The expected keys in each dict depends on the
+                List of dicts, such that `len(targets) == batch_size. The expected keys in each dict depends on the
```
Comment (alaradirik) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:2227]:
```suggestion
                List of dicts, such that `len(targets) == batch_size`. The expected keys in each dict depends on the
```

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py [hunk 6]
```diff
@@ -2172,6 +2160,39 @@ def loss_boxes(self, outputs, targets, indices, num_boxes):
...
+        """
+        if "pred_masks" not in outputs:
+            raise KeyError("No predicted masks found in outputs")
+
+        source_idx = self._get_source_permutation_idx(indices)
+        target_idx = self._get_target_permutation_idx(indices)
+        source_masks = outputs["pred_masks"]
+        source_masks = source_masks[source_idx]
+        masks = [t["masks"] for t in targets]
+        # TODO use valid to mask invalid areas due to padding in loss
```
Comment (alaradirik) [inline src/transformers/models/deformable_detr/modeling_deformable_detr.py:2177]: Is this left-over from this original codebase or will the invalid area computation be added?

**Review by alaradirik — APPROVED** [review]
LGTM! Just left a few comments and questions.

File: docs/source/en/model_doc/deformable_detr.mdx [hunk 7]
```diff
@@ -23,7 +23,7 @@ The abstract from the paper is the following:
 
 Tips:
 
-- One can use the [`AutoFeatureExtractor`] API to prepare images (and optional targets) for the model. This will instantiate a [`DetrFeatureExtractor`] behind the scenes.
+- One can use [`DeformableDetrFeatureExtractor`] to prepare images (and optional targets) for the model.
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply docs/source/en/model_doc/deformable_detr.mdx:26]: There's no `DeformableDetrProcessor`, do you mean `AutoProcessor`?

File: src/transformers/models/deformable_detr/modeling_deformable_detr.py (hunk 6 above)
Comment (NielsRogge) [reply src/transformers/models/deformable_detr/modeling_deformable_detr.py:2177]: This is a leftover from the original codebase.
